### DISCUSSION TITLE
Risk of corruption

### DISCUSSION THREAD

**Discussion by abique** [Serial-ATA/lofty-rs#352] [Q&A] [answered]
Hi,

I wonder how risky is it to save tags to an audio file.
What are the risks of corruption and what to do to reduce them?

Cheers,
Alex

**Comment by Serial-ATA** [comment, marked as answer]
Lofty has seen pretty wide adoption and over the years I've only gotten a few [reports of corruption](https://github.com/Serial-ATA/lofty-rs/issues?q=is%3Aissue+corrupt). Most cases come from storing **very** large items in Opus, which still doesn't seem to be *entirely* right. The best I can offer is to try it out and see how it works on your library, most people have no issues and all writes are at least *decently* tested for the most basic cases.

If you do encounter any issues with file corruption, just be sure to report it here. It'll very likely be one spurious file out of your entire library :smile:.

**Reply by abique** [reply]
I'm working on a DJ app and planning to store lightweight data:
- sample offset of first beat start
- loops
- hot cues
- tempo
- rating
- tags
- color
- album, artists, and some basic stuff

**Reply by abique** [reply]
I'll take your word and give it a try.

**Reply by Serial-ATA** [reply]
Assuming with data like that you'd be using ID3v2? In that case, it really shouldn't be an issue. They just get written at the start of the file anyway, so if Lofty messes up somewhere you can just delete the tag and your file should be fine. Full file corruption would only be an issue where the tags are embedded in the file somewhere.

**Reply by abique** [reply]
I'd be using anything that allows me to store non standard tags (I mean custom key/value pair).
I'm not enthusiast in storing data at the beginning of the file because what happens if there wasn't enough space reserved for the tag?

**Reply by abique** [reply]
Are tags stored as a linked list of tags? That way you have the file offset to the next tag and can append tags at the end of the file?

**Reply by Serial-ATA** [reply]
> I'm not enthusiast in storing data at the beginning of the file because what happens if there wasn't enough space reserved for the tag?

I'm not sure I understand what you mean by space reserved? Tags do not have a dedicated amount of space nor spot in the file most of the time. ID3v2 is meant to be written at the start of the file before any audio data.

> Are tags stored as a linked list of tags? That way you have the file offset to the next tag and can append tags at the end of the file?

Tags as represented in Lofty are purely a generic in-memory representation. No information about the file and its offsets are stored. That allows you to take a tag and write it to as many files as you'd like without having to worrying about the underlying file handling.

**Reply by abique** [reply]
So when you add new tags to a file, you may have to rewrite the entire file because it'd shift the audio content, is that correct?

**Reply by abique** [reply]
> > Are tags stored as a linked list of tags? That way you have the file offset to the next tag and can append tags at the end of the file?
> 
> Tags as represented in Lofty are purely a generic in-memory representation. No information about the file and its offsets are stored. That allows you to take a tag and write it to as many files as you'd like without having to worrying about the underlying file handling.

I was talking about the disk format.

**Reply by Serial-ATA** [reply]
Yes, unfortunately editing metadata will in many cases require an entire rewrite of the file. Some tag formats allow you to have padding, which Lofty uses to prevent rewrites where possible. That isn't an option in all formats, though.

> I was talking about the disk format.

On disk, the tag is ***usually*** just `[(size, item), (size, item)]`. We do not know ahead of time the number of items (most of the time) nor their validity until the parser reaches them. But that is highly dependent on the underlying format.

**Reply by abique** [reply]
When you perform a full rewrite, how do you do it?
Do you create a temporary file, write it, then swap it with the original?
Do you have a verification step before the swap?

**Reply by Serial-ATA** [reply]
1. The tag is created in memory (with necessary verification done during this step)
2. The file is read into memory
3. The tag is replaced in the file in memory
4. Any offsets/sizes are updated throughout the file
5. The file is rewritten to the original location

Lofty is restricted to certain portions of the file when performing the writes, and it does quite a bit of validation on your data before even creating the tag. It is, by design, much more aligned with the specs than other tagging libraries to try and prevent you from shooting yourself in the foot.

**Reply by abique** [reply]
Thank you for the details.
So if the application crashes in another thread during 5., then the file might be lost is that right?

It isn't a big issue actually. If I'd want to be extra safe, I could run the write operation in a child process do it on a copy of the original, then do the rename + unlink after some verification: read the metadata and make sure they are what I expect and hash the decoded audio and make sure it is the same as before.

**Reply by Serial-ATA** [reply]
Yeah, if for some reason something were to go wrong while it's writing then the file wouldn't be recoverable. That's why we try to do every bit of processing we can beforehand to minimize that risk.

**Reply by abique** [reply]
Thank you very much, this discussion was very helpful :-)

**Comment by uklotzde** [comment]
It is always the responsibility of the application to decide into which file metadata is written back. This could either be the original file or a temporary file. It is also the responsibility of the application to decide if and how to apply post-write validation/verification before replacing the original file.

Ideally, _Lofty_ must never leave behind a corrupt file. That's the expectation. But it would be naive to assume that this won't happen (again) in the future. Lofty is still young. Nevertheless, I consider it the best in class for Rust.

TagLib also suffered from file corruption bugs in the past.
